### PR TITLE
Document how to use the sign_out endpoint

### LINKED ISSUE #441: Document how to sign out
https://pusher.github.io/oauth2_proxy/endpoints mentions `/oauth2/sign_out` but does not explain how to use it. Neither does it mention whether it integrates somehow with the `end_session_endpoint` from https://my-oidc-provider.example.com/realms/mine/.well-known/openid-configuration

## Expected Behavior

I am trying to log out via https://my-app.example.com/oauth2/sign_out?rd=https%3A%2F%my-oidc-provider.example.com%2Frealms%2Fmine%2Fprotocol%2Fopenid-connect%2Flogout%3Fredirect_uri%3Dhttps%3A%2F%2Fexample.com%2F

expecting that 1. oauth2 will remove its cookies, 2. forward me to my OIDC provider logout url, 3. the provider will redirect  its cookies and redirect me to example.com

## Current Behavior

I arrive to https://my-app.example.com/index.html, obviously not logged out either from oauth2 or the OIDC provider.

## Implementation

1. Mention that the endpoint only removes the proxy's cookies and that you need to follow up with sign out with the oauth provider, for example by supplying a redirect parameter to be sent to the provider's sign out, e.g. _<domain>/oauth2/sign_out?rd=https%3A%2F%2Fmy-auth2-provider.example.com%2Frealms%2Fexample%2Fprotocol%2Fopenid-connect%2Flogout%3Fredirect_uri%3Dhttps%3A%2F%2Felg.no%2F_
2. The docs for sign out should mention that the domain to redirect to afterwards (via `?rd=`) needs to be included in `-whitelist-domain`. And the docs for `-whitelist-domain` should be changed from "after authentication" to "after authentication or sign out".

## Context

I don't know how to log out the users of my app either from the app or both the app and the OIDC provider.

## Environment

- Version used: v5

### PR DESCRIPTION
**Pull request by holyjak**
Fixes #441

+ a minor link fix

### CODE REVIEW

File: docs/5_endpoints.md [hunk 1]
```diff
@@ -17,3 +17,20 @@ OAuth2 Proxy responds directly to the following endpoints. All other endpoints w
 - /oauth2/callback - the URL used at the end of the OAuth cycle. The oauth app will be configured with this as the callback url.
 - /oauth2/userinfo - the URL is used to return user's email from the session in JSON format.
 - /oauth2/auth - only returns a 202 Accepted response or a 401 Unauthorized response; for use with the [Nginx `auth_request` directive](#nginx-auth-request)
+
+### Sign out
+
+To sign the user out, redirect him to `/oauth2/sign_out`. However, this endpoint only removes oauth2_proxy's own cookies, i.e. the user is still logged in with the OIDC provider and will automatically re-login when accessing the application again. What you want to do is to redirect him to the OIDC provider's sign out page afterwards using the `rd` query parameter, i.e. redirect the user to something like (notice the url-encoding!):
```
Comment (JoelSpeed) [inline docs/5_endpoints.md]:
```suggestion
To sign the user out, redirect them to `/oauth2/sign_out`. This endpoint only removes oauth2_proxy's own cookies, i.e. the user is still logged in with the authentication provider and may automatically re-login when accessing the application again. You will also need to redirect the user to the authentication provider's sign out page afterwards using the `rd` query parameter, i.e. redirect the user to something like (notice the url-encoding!):
```

File: docs/5_endpoints.md [hunk 2]
```diff
@@ -17,3 +17,20 @@ OAuth2 Proxy responds directly to the following endpoints. All other endpoints w
...
+
+    /oauth2/sign_out?rd=https%3A%2F%2Fmy-oidc-provider.example.com%2Fsign_out_page
```
Comment (JoelSpeed) [inline docs/5_endpoints.md]: This should be in a code block

File: docs/5_endpoints.md [hunk 3]
```diff
@@ -17,3 +17,20 @@ OAuth2 Proxy responds directly to the following endpoints. All other endpoints w
...
+
+Alternatively, include the redirect URL in the `X-Auth-Request-Redirect` header:
+
+    GET /oauth2/sign_out HTTP/1.1
+    X-Auth-Request-Redirect: https://my-oidc-provider/sign_out_page
+    ...
```
Comment (JoelSpeed) [inline docs/5_endpoints.md]: This should be in a code block

**Comment by holyjak** [comment]
I think that would be great. Nr 1 - provider first - would require that
there was a standard redirect param. Our provider supports `redirect_uri`
but I have no idea whether it is common.

Option 2 - the proxy redirects automatically to the end session url from
the Metadata (and add it to the whitelist) - again you likely want to
redirect somewhere after the final sign out and thus need to provide the
redirect uri to the provider so we would need to support also something
like `-signout-redirect-params` (where I would set
`-signout-redirect-params redirect_uri=http://elg.no`

File: docs/5_endpoints.md (hunk 2 above)
Comment (holyjak) [reply docs/5_endpoints.md]: It is indented by 4 spaces and thus is a code block. Or did you mean "fenced code block"?

**Comment by Ornias1993** [comment]
@holyjak A bit of a necro, but if someone manages to keep the cookie (for example a user on a malicieus PC). Does this mean they could use said cookie to pass the proxy even though they are still logged out from the IDP?

If so, would using redis prevent this?

**Pull request merged**